### PR TITLE
cache the return of Scan.grad

### PR DESCRIPTION
**Pull request by kyunghyuncho**
If an option "cache_grad" is set to True when calling theano.scan, the result of Scan.grad will be cached. The default value of "cache_grad" is False so that there will be no issue with the existing code.

### CODE REVIEW

File: theano/scan_module/scan.py [hunk 1]
```diff
@@ -990,6 +992,10 @@ def wrap_into_list(x):
     info['as_while'] = as_while
     info['profile'] = profile
     info['allow_gc'] = allow_gc
+    info['cache_grad'] = cache_grad
+    if cache_grad:
+        warnings.warn('cache_grad works only when Scan Op is associated'
```
Comment (nouiz) [inline theano/scan_module/scan.py]: I would make this warning more threatening. Something like:

Are are using Scan cache_grad. This is dangerous and can lead to WRONG results as it only work in some condition. This include that it work only with a single cost. If all conditions aren't meet, no error are raised, but wrong result are returned!

Comment (nouiz) [inline theano/scan_module/scan.py]: Is there other condition? I prefer to let it vague, just in case we forget some condition.

**Comment by kyunghyuncho** [comment]
Made the warning message more serious.

**Comment by nouiz** [comment]
I let @abergeron, @lamblin and/or @carriepl decide if we should include
this. I do not have other comment, but I do not know enough that part of
scan.

**Comment by abergeron** [comment]
This seems to be more in the category of "dirty hacks to go faster" than
"feature that should go in theano".  I could be swayed otherwise though.

2015-03-13 20:15 GMT-04:00 Frédéric Bastien <notifications@github.com>:

**Comment by kyunghyuncho** [comment]
Without this, it takes more than several hours for some models that involve
many Scan Ops to do tensor.grad.

**Comment by nouiz** [comment]
I agree, it seem more like a hack. @lamblin and @carriepl have a plan for a
good fix that can be done by default. But until this is done, I think a
hack isn't too bad due to the consequence (hours for theano.grad)

**Comment by abergeron** [comment]
I'm still not convinced we should merge this, especially if there is an alternate solution in the works.  But that shouldn't stop you from using this code with problematic models if you need it right now.

**Comment by kyunghyuncho** [comment]
Can Pascal or Pierre-Luc explain to me a bit of what an alternative
solution be like?

**Comment by lamblin** [comment]
The alternative solution would be to go through the inner graph, similarly to what `theano.grad` does, piecing together the global connection pattern, rather than calling `theano.grad` explicitly.
I don't know how much work that would be, or how long it would take, though.

**Comment by kyunghyuncho** [comment]
Then, how about we merge this PR *after* the release of 0.7? This will help
us not adding any hackish solution to a stable release, but give people a
solution to avoid the exploding computation in tensor.grad until the
alternative one is implemented.

**Comment by kyunghyuncho** [comment]
I narrowed down the underlying cause of this speed issue involving Scan Op. I will make a new PR.

**Pull request closed without merging**